### DISCUSSION TITLE
EccLevel::fromString()

### DISCUSSION THREAD

**Discussion by raubv0gel** [chillerlan/php-qrcode#160] [Ideas]
For `\chillerlan\QRCode\Common\EccLevel` a static function like `fromString()` would be nice. I. e., `EccLevel::fromString('H')` should return `EccLevel::H` and so on. Lower case and upper case letter should be possible.

**Comment by codemasher** [comment]
Hmm, can you provide me with a use case that is commonly viable? The current implementation is to serve all the internal needs (before you ask: the `__toString()` method is for debugging/testing only). If you want to access the classes constants from a variable, you can just do this:
```php
$var = 'h';
$eccLevel = constant(EccLevel::class.'::'.strtoupper($var));
```

**Reply by raubv0gel** [reply]
Nice suggestion! Use case is: we have a config `QR.ECC` which is a string (like `H`, `h`). And I want to create an EccLevel from this `QR.ECC`.

```php
$var = 'h';
$eccLevel = constant(EccLevel::class.'::'.strtoupper($var));
```
but I’m not sure if this is the „perfect way“. There should be exception handling (what if `$var === 'foo'`?). And what if the Implementation of `EccLevel` changes – let’s say `EccLevel::H` becomes `EccLevel::High` in future?

I can try to code an implementation and make a PR if you like …

**Reply by codemasher** [reply]
See, what i mean by a "viable" use case is that it is useful to and can be re-used by many projects. However, setting the ECC level (or any config value) from variables is just not as common or useful as you might think. One specific use case would be e.g. an [online QR Code generator](https://github.com/chillerlan/php-qrcode/blob/c03aafcf183e297128931c40fb425fb40057ea86/examples/qrcode-interactive.php#L59) (which i personally despise for security and privacy reasons) or as you suggest from an external config. From what i've seen in code [from dependents and implementors](https://github.com/chillerlan/php-qrcode/network/dependents) even this is an uncommon use case as QR Codes are usually used for a specific task and design with either fixed settings or just with the plain defaults.

I guess the "perfect" way for your use case wold be like 5 extra lines in your code - this would be at least how i'd implement such a method:
```php
$ecc = strtoupper($ecc_value_from_config);

if(!in_array($ecc, ['L', 'M', 'Q', 'H'])){
    throw new Exception('invalid ecc level');
}

$qrOptions->eccLevel = constant(EccLevel::class.'::'.$ecc);
```
On a side note: a method `EccLevel::fromString()` would rather return an `EccLevel` instance rather than just the value that can be accessed via a public constant anyway.

**Reply by codemasher** [reply]
Thinking about it - what i can offer to do is to change [the ECC level setter](https://github.com/chillerlan/php-qrcode/blob/c03aafcf183e297128931c40fb425fb40057ea86/src/QROptionsTrait.php#L391-L398) so that it accepts `int|string`, but this won't happen any time soon, as v5 will still support PHP 7.4.

**Reply by codemasher** [reply]
So if you can't wait and you're running PHP 8 or higher, you can do this:
```php
use chillerlan\Settings\SettingsContainerAbstract;
use chillerlan\QRCode{QRCodeException, QROptionsTrait};
use chillerlan\QRCode\Common\EccLevel;
use function constant, in_array, is_string, sprintf, strtoupper;

class MyQROptions extends SettingsContainerAbstract{
	use QROptionsTrait{
		set_eccLevel as private set_eccLevel_old; // we need to rename the method to not cause a type hint conflict
	}

	protected function set_eccLevel(int|string $eccLevel):void{

		if(is_string($eccLevel)){
			$ecc = strtoupper($eccLevel);

			if(!in_array($ecc, ['L', 'M', 'Q', 'H'])){
				throw new QRCodeException(sprintf('invalid ecc level: %s', $ecc));
			}

			$eccLevel = constant(EccLevel::class.'::'.$ecc);
		}

		if((0b11 & $eccLevel) !== $eccLevel){
			throw new QRCodeException(sprintf('Invalid error correct level: %s', $eccLevel));
		}

		$this->eccLevel = $eccLevel;
	}
}
```
~~Edit: i think it might complain about the type hint when done this way but i think you get the gist of it.~~ fixed.

**Reply by codemasher** [reply]
I've implemented this now in the v6 dev branch (not yet public), that is (currently) intended for PHP 8.2+.

**Comment by codemasher** [comment]
Just adding the note here that as of PHP 8.3, class constants can be called from a variable:

```php
$eccLevel = EccLevel::{$var};
```

See https://wiki.php.net/rfc/dynamic_class_constant_fetch
